### PR TITLE
PDF styles.

### LINKED ISSUE #2563: Style print/PDF version of Area of Study page.
# Story

As an Admissions staff member I want to be able to print (or download a PDF of) a styled version of Areas of Study pages so that they look beautiful when they are sent to a prospective student.

# Proposed solution

Based on the results of #2500, implement some way of outputting specific fields in a specific format for an Area of Study page when viewed as a printable page or PDF.

Todo:

- [x] Remove 'View PDF' link shown for authenticated users at the bottom of AOS pages.
- [x] Get TTF fonts working in PDF.
- ~~Create specific role for access AOS PDFs and listing view.~~
  - Authenticated role has permission. Webmasters can add users with viewer role for most-restrictive access.
- [x] Recreate view of AOS print links without print instructions: https://admissions.uiowa.edu/at-iowa-print
- [x] Set downloaded filename to mail item code field.
- [x] Refactor imported UIDS partials as PDF-local versions to avoid generation failure.
- [x] Create on-brand PDF with an existing template in mind: https://brand.uiowa.edu/template-library

### PR DESCRIPTION
**Pull request by richardbporter**
Resolves #2563 

### Testing
- `yarn workspace admissions_core run build`
- `blt sync --site admissions.uiowa.edu`
- Go to https://admissions.local.drupal.uiowa.edu/areas-of-study/print-links and ensure access is denied while logged out.
- Log in and revisit link above as viewer or other role and ensure access granted.
- Click Accounting from the print links view page and ensure PDF viewer renders.
- Edit the Accounting AOS and add a value to the Mail Code field.
  - http://admissions.local.drupal.uiowa.edu/areas-of-study/accounting
- Go back to the PDF viewer for accounting and download PDF and ensure filename is set to the Mail Code value entered.
  - http://admissions.local.drupal.uiowa.edu/print/pdf/node/761 
- Review PDF display, styles and templates. See #2970 for instructions on working with SASS/template changes.
  - http://admissions.local.drupal.uiowa.edu/print/pdf/node/761/debug
  - Add a really long page title and test that it wraps within the pdf

### CODE REVIEW

**Comment by richardbporter** [comment]
@bspeare I'm not having any luck with this. 😢

**Comment by bspeare** [comment]
@richardbporter `body  {font-family: 'Zilla Slab';}` worked on everything but that page title.

**Comment by richardbporter** [comment]
I'm not sure I understand this comment but seems to be related to font-variant somehow: https://www.drupal.org/project/entity_print/issues/2806763#comment-11753902

**Comment by richardbporter** [comment]
I think I have some fonts working. Obviously, more adjustments are needed but they should be loading at least.

**Comment by richardbporter** [comment]
Can someone with Adobe CC installed help with getting the the logo in the PDFs? We can't use SVG here, just an `img` element and I think it needs to be cropped. Maybe @bspeare or @GaryRidgway ?

**Comment by richardbporter** [comment]
The fontawesome library is currently breaking PDF generation so we'll need to find another way to do that. I'm also having trouble getting the clearfix utlitity to work. Maybe that wouldn't be necessary if there was another to do the columns - I'm using float.

**Comment by bspeare** [comment]
@richardbporter I'll take a look.

**Comment by richardbporter** [comment]
Todo:

- [x] Add envelope and telephone icons as images.
- [x] Replace `i` elements under learn more with icon images.
- [x] Review fonts, font sizing, spacing, etc. and adjust to convey brand voice.
- [x] Style `Accounting (BBA)` inline with parentheses. 
- [x] Get Antonio font to render with button. 
- [x] Get Majors to appear inline and separated by commas. 
Screen Shot 2021-02-15 at 8 35 34 AM

- [x] Add `field_area_of_study_teaching`, `field_area_of_study_teaching_txt` , `field_area_of_study_honors`, `field_area_of_study_four_year`, `field_area_of_study_four_txt`, `field_area_of_study_selective`, `field_area_of_study_select_txt`, `field_area_of_study_competitive`, `field_area_of_study_comp_txt` to appear inline and separated by commas below "Minors". 
Screen Shot 2021-02-15 at 8 34 44 AM

**Comment by richardbporter** [comment]
I'm going to switch to D9 testing for a bit if you want to adjust any styles here @bspeare .

**Comment by richardbporter** [comment]
@bspeare Anything specific I can help with to get this over the finish line? Are the checkboxes above up to date?

**Comment by bspeare** [comment]
@richardbporter I just updated the checkboxes with a few new ones if you want to attempt one of those items.

**Comment by bspeare** [comment]
@richardbporter I started working on the "Get Majors to appear inline and separated by commas" checkbox on Friday, and I can continue on that one this morning.

**Comment by bspeare** [comment]
@richardbporter I think this is ready for review if you want to check the changes.

**Comment by richardbporter** [comment]
I think it looks great. Was there another AOS with different data to test with? I'm still using Accounting. 

I made a change to how ignored assets are handled which hopefully will make it easier to add fonts, images, etc. to assets directories going forward. We had to force-add things like that in the past which was not great.

Otherwise, I'd mark this for review for someone else to look over and then ship it.

**Comment by bspeare** [comment]
@richardbporter We don't have a great full example to test with besides Accounting.  For my local testing, I added all fields and as many multiple value fields where applicable to try and stress test the layout.

**Comment by richardbporter** [comment]
Are we using the arrow.png, envelope.png and phone.png as well? Or are just the gray versions in use?

**Comment by bspeare** [comment]
@richardbporter I think arrow.png is used in the top button.   I think we can leave out the envelope.png and phone.png and use gray only.

**Comment by richardbporter** [comment]
The bottom margin for majors/minors seems larger than the other multi-value degree-type fields. Also, the commas don't seem quite right, at least on my local environment. I recompiled and CR'ed but I'm seeing this:

image

**Comment by bspeare** [comment]
@richardbporter I'll take another look.  That is not the intended display.

**Comment by bspeare** [comment]
Here's what it looks like with a fresh `blt ds`:

Screen Shot 2021-02-16 at 8 44 45 AM

**Comment by joewhitsitt** [comment]
I want to major in cookies and minor in running. you know, to offset the cookies

**Comment by richardbporter** [comment]
Ok cool, I see it now. Looks good to me.

**Comment by richardbporter** [comment]
Unless @bspeare has any more work planned for this, I would say it'd be good for someone besides us to look it over.

**Comment by bspeare** [comment]
No work planned currently, ready for review.

**Comment by joewhitsitt** [comment]
pulling down for testing

**Comment by joewhitsitt** [comment]
Logged in, I still get access denied on https://admissions.local.drupal.uiowa.edu/areas-of-study/print-links. had to go to http://admissions.local.drupal.uiowa.edu/areas-of-study/print-links

mail code stripped my exclamation point from `joerocks!`

**Review by joewhitsitt — APPROVED** [review]
Test steps check out and I wasn't able to get into trouble. Did not account for all fields being represented in the pdf.

**Comment by richardbporter** [comment]
> Logged in, I still get access denied on https://admissions.local.drupal.uiowa.edu/areas-of-study/print-links. had to go to http://admissions.local.drupal.uiowa.edu/areas-of-study/print-links
> 
> mail code stripped my exclamation point from `joerocks!`

I'm confused about this.

**Comment by joewhitsitt** [comment]
if i typed joerocks! in the field, the filename ended up being joerocks, but on the form it is still joerocks!

my local http/https made the test instructions hard to follow

**Comment by richardbporter** [comment]
> if i typed joerocks! in the field, the filename ended up being joerocks, but on the form it is still joerocks!
> 
> my local http/https made the test instructions hard to follow

Ok. Do you think we should recreate as a number field?

**Comment by joewhitsitt** [comment]
help text that says special characters and spaces will be stripped?

**Comment by richardbporter** [comment]
Thanks all!
